File: backend/chat-store-service/src/main/java/com/example/chatstoreservice/message/ChatKafkaMessageListener.java
package com.example.chatstoreservice.message;

import com.example.chatstoreservice.chat.entity.Chat;
import com.example.chatstoreservice.chat.service.ChatService;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.kafka.support.Acknowledgment;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class ChatKafkaMessageListener {

    private final ChatService chatService;
    private final ObjectMapper objectMapper;

    @KafkaListener(topics = "modu.modu-chat.chat", containerFactory = "chatKafkaListener", groupId = "chat-store")
    public void debeziumListener(ConsumerRecord<String, ChatListener> consumerRecord, Acknowledgment acknowledgment) {
        try {
            ChatListener chatListener = consumerRecord.value();
            String op = chatListener.getPayload().getOp();
            ChatModel chatModel = chatListener.getPayload().getAfter();

            Chat chat = new Chat(chatModel);

            if (op.equals("c")) {
                chatService.createChat(chat);
            } else if (op.equals("u")) {
                chatService.updateChat(chat);
            } else if (op.equals("d")) {
                chatService.deleteChat(chat);
            } else {
                // do nothing
            }

            log.info(chatListener.toString());
        } catch (Exception e) {
            log.error(e.getMessage());
        }

        acknowledgment.acknowledge();
    }
}

File: backend/chat-store-service/src/main/kotlin/com/example/chatstoreservice/message/ChatKafkaMessageListener.kt
package com.example.chatstoreservice.message

import com.example.chatstoreservice.chat.entity.Chat
import com.example.chatstoreservice.chat.service.ChatService
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.slf4j.LoggerFactory
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.support.Acknowledgment
import org.springframework.stereotype.Service

/** Debezium 이 MySQL chat 테이블 변경을 실어 보내는 토픽을 받아 Mongo 미러에 반영한다. */
@Service
class ChatKafkaMessageListener(private val chatService: ChatService) {

    private val log = LoggerFactory.getLogger(ChatKafkaMessageListener::class.java)

    @KafkaListener(topics = ["modu.modu-chat.chat"], containerFactory = "chatKafkaListener", groupId = "chat-store")
    fun debeziumListener(consumerRecord: ConsumerRecord<String, ChatListener>, acknowledgment: Acknowledgment) {
        try {
            val chatListener = consumerRecord.value()
            val payload = chatListener.payload
            val chatModel = payload?.after
            if (payload != null && chatModel != null) {
                val chat = Chat(chatModel)
                when (payload.op) {
                    "c" -> chatService.createChat(chat)
                    "u" -> chatService.updateChat(chat)
                    "d" -> chatService.deleteChat(chat)
                    else -> Unit
                }
            }
            log.info(chatListener.toString())
        } catch (e: Exception) {
            log.error(e.message)
        }
        acknowledgment.acknowledge()
    }
}
